state/state-network.md

```diff
@@ -305,46 +308,44 @@ state root and prove that *"target"* node is part of the trie defined by that st
+- all contract bytecode for newly created contracts
+    - proof has to include the proof for the account trie that corresponds to the same contract
 
-> For example, let this be the proof: `[A, B, C, D, E, F]`, then `A` is the root note and `F` is
-> the target node. This proof should be gossiped to nodes that are interested in storing node `F`.
+A bridge should compute the content-id values for all content key/value pairs. These content-ids
+should be sorted by proximity to its own node-id. Beginning with the content that is *closest* to its
+own node-id it should proceed to GOSSIP each individual content to nodes *interested* in that content.
 
-The recipients of this gossip are then responsible for gossiping the penultimate node of the proof
-(parent of the target node, `E` in our example). To do so, they would strip the last (target) node
-from the proof, create content-key with `E` being the target node, search the DHT for nodes that
-are interested in newly created content-key and gossip the content with proof to them.
+#### Special case to consider
 
-> It should be highlighted that creation of the content-key also requires updating the `path` field
-by removing one or more (in case of an extension node) nibbles from the end.
+It should be highlighted that it's possible for a trie node to be modified even if it is not in the
+path of any value that was modified in the block. Consider following example:
 
-This process repeats until it terminates at the state root, with the final round of gossip only
-containing the `[A]` which is the state root node of the trie.
+Let's assume that trie contains two key/value pairs: `(0x0123, A)` and `(0x1234, B)`. Trie would look
+something like this (numbers next to branches indicate the index in the branch node):
 
-### Bridge Node Responsibilities
+```
+         branch (root)
+       /0             \1 
+prefix: 123            prefix: 234
+value:  A              value:  B
+```
 
-The bridge is responsible for creating and gossiping all following data and their proofs:
+New key/value `(0x0246, C)` is inserted in the next block, resulting in the new trie:
 
-- account trie data:
-    - all of the new and modified account nodes from the state trie
-- contract storage trie data:
-    - all of the new and modified storage slots from each modified contract storage trie
-    - proof has to include the proof for the account trie that corresponds to the same contract
-    - recursive gossip stops at the root of the storage trie
-- all contract bytecode for newly created contracts
-    - proof has to include the proof for the account trie that corresponds to the same contract
-    - recursive gossip doesn't happen in this case
+```
+                   branch (root)
+                /0              \1
+          branch                 prefix: 234
+        /1      \2               value:  B
+prefix: 23       prefix: 46
+value:  A        value:  C
+```
```

Comment:
@morph-dev Interesting example. Originally I was thinking that it would be enough to gossip each updated leaf (with parent nodes/proof) in order to get the correct state for each block. 

I guess this means the bridge nodes will need to somehow detect which nodes have changed for each block by looking through the trie. 
Is there a suggested way we could find out which nodes have changed that are not in the path of the updated leafs? 

Assuming the state bridge executes each block of transactions locally, it could loop through the nodes in the trie and compare the state trie before and after executing the block to determine which nodes changed. There are probably more efficient solutions though.

Reply:
I believe the only known way to do this is to monitor at the trie library level for changes.

Reply:
I would say that it depends a lot on the exact implementation of these components, as I'm not sure if there is single solution for all of them.

As Piper said, one approach is to monitor changes at the trie library level.

Another approach that I believe would work correctly with most implementations is to monitor trie nodes writes to the DB (I assume that each updated trie node has to be written to the db at some point).